### ISSUE TITLE
Frame as RL problem

### ISSUE THREAD

**Issue by LiamConnell** [LiamConnell/DrawAI#2]
Control pen with velocity and angular acceleration (or just acceleration). 

One reward is kl
Divergence of v and a to normal.

One is similarity or likeness to image.

How to define likeness?

Shading?

Lines? Borders?

Recognizability as category?

**Comment by LiamConnell** [comment]
### Design Proposal: Reframing Drawing as a Reinforcement Learning Problem

Here is a detailed design proposal for implementing the drawing process as a Reinforcement Learning (RL) task, based on the ideas in this issue. This approach shifts from the current deterministic pipeline to a dynamic, learning-based agent.

---

#### **1. Core RL Components**

*   **Environment ()**: This will be a custom Gym-like environment.
    *   **Reference Image**: The input image, pre-processed to extract an edge map (e.g., via Canny) and a saliency map. These will be crucial for the reward function.
    *   **Canvas**: A digital canvas (e.g., a NumPy array) where the agent's drawing is rendered.
    *   **State ()**: The state at time  will be a vector containing:
        *   **Pen State**:  (position, velocity, and pen pressure/state).
        *   **Canvas State**: A compressed representation of the current drawing. Instead of the full pixel buffer, this could be a feature vector from a simple CNN applied to the canvas, or a history of recent pen states.
        *   **(Optional) Local Image Patch**: A small patch from the reference image's edge/saliency map around the current pen position to give the agent local context.
    *   ****: Updates the environment by applying the agent's action, calculating the reward, and returning the next state.
    *   ****: Resets the environment to an initial state for a new episode.

*   **Agent**: A standard RL agent from a library like . An Actor-Critic model like **PPO (Proximal Policy Optimization)** or **SAC (Soft Actor-Critic)** would be a good choice, as they work well with continuous action spaces.

*   **Action Space**: A continuous action space seems most natural.
    *   **Option A (Velocity Control)**: The action  is a 2D vector  defining the pen's velocity. This is simpler to start with.
    *   **Option B (Acceleration Control)**: The action  is a 2D vector . This allows for smoother, more physically realistic curves but adds complexity.

*   **Reward Function ()**: This is the most critical component and will be a weighted sum of several factors:
    *   ** (Image Similarity)**: The primary reward for matching the reference image.
        *   **Initial Approach**: Reward the agent for drawing on high-value areas of the pre-computed edge or saliency maps. For a new line segment, the reward is the sum of the pixel values it covers on the map.
        *   **Advanced Approach**: Use a perceptual loss function. Extract features from the current canvas and the reference image using a pre-trained VGG network. The reward is the negative L2 distance between these feature vectors. This is computationally intensive but captures high-level similarity better.
    *   ** (Exploration Bonus)**: To prevent the agent from getting stuck in one area, provide a small bonus for drawing over previously unvisited parts of the edge/saliency map.
    *   ** (Path Regularization)**: Penalize the agent for high curvature or drastic changes in acceleration to encourage smoother lines. This directly addresses a point from the issue.
    *   ** (Penalty)**: A large negative reward if the pen goes off the canvas.
    *   **Total Reward**: . These weights will require careful tuning.

---

#### **2. Implementation Plan**

1.  **Environment First**: Implement the  class. This is the foundation.
2.  **Integrate a Framework**: Use  to provide the PPO or SAC algorithm.
3.  **Define Policy Network**: A simple Multi-Layer Perceptron (MLP) should suffice for the policy network (mapping state to action).
4.  **Iterate on Rewards**: Start with the simplest reward components ( based on edge map, ) and incrementally add more complexity.
5.  **Training**: Train the agent, periodically saving the model and generating sample drawings to evaluate progress.

---

#### **3. Open Questions & Trade-offs**

*   **State Representation**: How complex does the state need to be? A smaller state space (e.g., only recent pen history instead of the full canvas) will train much faster.
*   **Reward vs. Performance**: The perceptual loss for  is powerful but will dramatically slow down training. The edge-map approach is much faster. We should start with the faster approach.
*   **Enforcing 'One-Line'**: The simplest way to enforce the one-line constraint is to not give the agent a 'pen up' action. The episode ends when a time limit is reached or the agent goes off-screen.

This RL approach is a significant but exciting departure from the current method. It trades the predictability of a deterministic algorithm for the potential of a more adaptive and expressive drawing agent.

**Comment by LiamConnell** [comment]
# Proposed RL Solution for AI Drawing

## Problem Formulation

**Objective**: Train an RL agent to recreate target images through sequential pen strokes, balancing quality, efficiency, and natural drawing behavior.

## Environment Design

**State Space**:
- Current canvas state (H×W×C tensor)
- Current pen position/orientation (x, y, θ)
- Target image (H×W×C tensor) 
- Stroke history/context (optional)

**Action Space**:
- Pen movement: (Δx, Δy) velocity components
- Pen pressure/width: continuous value for stroke thickness
- Pen state: up/down (discrete) for stroke segmentation

## Multi-Objective Reward Function

### 1. Perceptual Similarity Reward (0.4 weight)
```
R_perceptual = -LPIPS(canvas, target) - λ₁ * MSE(canvas, target)
```
- LPIPS for perceptual quality
- MSE for pixel-level accuracy
- Consider CLIP embeddings for semantic similarity

### 2. Stroke Quality Reward (0.3 weight)
```
R_stroke = -KL(velocity_dist || natural_velocity_dist) - λ₂ * jerk_penalty
```
- KL divergence to encourage natural stroke velocities
- Jerk penalty (derivative of acceleration) for smoothness
- Curvature consistency for realistic pen movements

### 3. Efficiency Reward (0.2 weight)
```
R_efficiency = -α * total_path_length - β * num_pen_lifts
```
- Minimize total drawing distance
- Penalize excessive pen-up movements
- Encourage economical stroke planning

### 4. Edge/Structure Reward (0.1 weight)
```
R_structure = IoU(canny(canvas), canny(target))
```
- Canny edge detection similarity
- Captures structural/geometric accuracy

## Scalability Requirements

**Vectorized Environment**:
- JAX-based implementation for GPU parallelization
- Differentiable canvas rendering (consider `skia-python` or custom CUDA kernels)
- Batch size: 1000+ parallel environments

**Efficient Reward Computation**:
- Pre-compute target image features (LPIPS, edges, CLIP embeddings)
- GPU-accelerated similarity metrics
- Shared memory for target images across environments

**Technical Stack**:
- JAX for environment vectorization
- Flax/Haiku for policy networks
- Differentiable rendering backend
- PPO/SAC with multi-objective optimization

## Training Strategy

1. **Curriculum Learning**: Start with simple shapes, progress to complex images
2. **Multi-Task Training**: Mix different image types/styles
3. **Reward Shaping**: Gradually shift from pixel-level to perceptual rewards
4. **Behavioral Cloning**: Pre-train on human drawing demonstrations if available

This approach balances artistic quality with computational efficiency while maintaining scalability for large-scale training.
